man/tpm2_readpublic.1.md
% tpm2_readpublic(1) tpm2-tools | General Commands Manual

# NAME

**tpm2_readpublic**(1) - Read the public area of a loaded object.

# SYNOPSIS

**tpm2_readpublic** [*OPTIONS*]

# DESCRIPTION

**tpm2_readpublic**(1) - Reads the public area of a loaded object.

# OPTIONS

  * **-c**, **\--object-context**=_OBJECT_:

    Context object for the object to read.

  * **-n**, **\--name**=_FILE_:

    An optional file to save the name structure of the object.

[pubkey options](common/pubkey.md)

    Public key format.

  * **-o**, **\--output**=_FILE_:

    The output file path, recording the public portion of the object.

  * **-t**, **\--serialized-handle**=_HANDLE_:

    If the object to be read is a persistent object specified by a raw handle,
    optionally save the serialized handle for use later. This routine does NOT verify the name of the object being read. Callers should ensure that the
    contents of name match the expected objects name.

  * **-q**, **\--qualified-name**=_FILE_:

    Saves the qualified name of the object to _FILE_. The qualified name of the object is the name algorithm hash of
    the parents qualified and the objects name. Thus the qualified name of the object serves as proof of the objects
    parents.

## References

[context object format](common/ctxobj.md) details the methods for specifying
_OBJECT_.

[common options](common/options.md) collection of common options that provide
information many users may expect.

[common tcti options](common/tcti.md) collection of options used to configure
the various known TCTI modules.
# EXAMPLES

## Create a primary object and read the public structure in an openssl compliant format
```bash
tpm2_createprimary -c primary.ctx
tpm2_readpublic -c primary.ctx -o output.dat -f pem
```

## Serialize an existing persistent object handle to disk for later use

This work-flow is primarily intended for existing persistent TPM objects. This
work-flow does not verify that the name of the serialized object matches the
expected, and thus the serialized handle could be pointing to an attacker
controlled object if no verification is done. If you are creating an object from
scratch, save the serialized handle when making the object persistent.

We assume that an object has already been persisted, for example via:

```bash
# We assume that an object has already been persisted, for example
tpm2_createprimary -c primary.ctx

# context files have all the information for the TPM to verify the object
tpm2_evictcontrol -c primary.ctx
persistent-handle: 0x81000001
action: persisted
```

Next use the persistent handle to get a serialized handle:

```bash
# The persistent handle output could be at an attacker controlled object,
# best practice is to use the option "-o: for tpm2_evictcontrol to get a
# serialized handle instead.

tpm2_readpublic -c 0x81000001 -o output.dat -f pem -t primary.handle

# use this verified handle in an encrypted session with the tpm
tpm2_startauthsession --policy-session -S session.ctx -c primary.handle
```

For new objects, its best to use all serialized handles.

[returns](common/returns.md)

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/tpm2_readpublic.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <stdbool.h>
#include <stdlib.h>

#include "files.h"
#include "log.h"
#include "tpm2.h"
#include "tpm2_convert.h"
#include "tpm2_tool.h"

typedef struct tpm_readpub_ctx tpm_readpub_ctx;
struct tpm_readpub_ctx {
    struct {
        UINT8 f :1;
    } flags;
    char *output_path;
    char *out_name_file;
    char *out_qname_file;
    tpm2_convert_pubkey_fmt format;
    tpm2_loaded_object context_object;
    const char *context_arg;
    const char *out_tr_file;
};

static tpm_readpub_ctx ctx = {
    .format = pubkey_format_tss,
};

static tool_rc read_public_and_save(ESYS_CONTEXT *ectx) {

    TPM2B_PUBLIC *public;
    TPM2B_NAME *name;
    TPM2B_NAME *qualified_name;

    tool_rc rc = tool_rc_general_error;

    tool_rc tmp_rc = tpm2_readpublic(ectx, ctx.context_object.tr_handle,
            &public, &name, &qualified_name);
    if (tmp_rc != tool_rc_success) {
        return tmp_rc;
    }

    tpm2_tool_output("name: ");
    UINT16 i;
    for (i = 0; i < name->size; i++) {
        tpm2_tool_output("%02x", name->name[i]);
    }
    tpm2_tool_output("\n");

    bool ret = true;
    if (ctx.out_name_file) {
        ret = files_save_bytes_to_file(ctx.out_name_file, name->name,
                name->size);
        if (!ret) {
            LOG_ERR("Can not save object name file.");
            goto out;
        }
    }

    tpm2_tool_output("qualified name: ");
    for (i = 0; i < qualified_name->size; i++) {
        tpm2_tool_output("%02x", qualified_name->name[i]);
    }
    tpm2_tool_output("\n");

    tpm2_util_public_to_yaml(public, NULL);

    ret = ctx.output_path ?
            tpm2_convert_pubkey_save(public, ctx.format, ctx.output_path) :
            true;
    if (!ret) {
        goto out;
    }

    if (ctx.out_qname_file) {
        ret = files_save_bytes_to_file(ctx.out_qname_file, qualified_name->name,
                qualified_name->size);
        if (!ret) {
            goto out;
        }
    }

    if (ctx.out_tr_file) {
        rc = files_save_ESYS_TR(ectx, ctx.context_object.tr_handle,
                ctx.out_tr_file);
    } else {
        rc = tool_rc_success;
    }

out:
    free(public);
    free(name);
    free(qualified_name);

    return rc;
}

static bool on_option(char key, char *value) {

    switch (key) {
    case 'c':
        ctx.context_arg = value;
        break;
    case 'o':
        ctx.output_path = value;
        break;
    case 'f':
        ctx.format = tpm2_convert_pubkey_fmt_from_optarg(value);
        if (ctx.format == pubkey_format_err) {
            return false;
        }
        ctx.flags.f = 1;
        break;
    case 'n':
        ctx.out_name_file = value;
        break;
    case 't':
        ctx.out_tr_file = value;
        break;
    case 'q':
        ctx.out_qname_file = value;
        break;
    }

    return true;
}

bool tpm2_tool_onstart(tpm2_options **opts) {

    static const struct option topts[] = {
        { "output",            required_argument, NULL, 'o' },
        { "object-context",    required_argument, NULL, 'c' },
        { "format",            required_argument, NULL, 'f' },
        { "name",              required_argument, NULL, 'n' },
        { "serialized-handle", required_argument, NULL, 't' },
        { "qualified-name",    required_argument, NULL, 'q' }
    };

    *opts = tpm2_options_new("o:c:f:n:t:q:", ARRAY_LEN(topts), topts, on_option,
            NULL, 0);

    return *opts != NULL;
}

static tool_rc init(ESYS_CONTEXT *context) {

    tool_rc rc = tpm2_util_object_load(context, ctx.context_arg,
            &ctx.context_object, TPM2_HANDLE_ALL_W_NV);
    if (rc != tool_rc_success) {
        return rc;
    }

    bool is_persistent = ctx.context_object.handle
            && ((ctx.context_object.handle >> TPM2_HR_SHIFT)
                    == TPM2_HT_PERSISTENT);
    if (ctx.out_tr_file && !is_persistent) {
        LOG_ERR("Can only output a serialized handle for persistent object "
                "handles");
        return tool_rc_general_error;
    }

    return tool_rc_success;
}

tool_rc tpm2_tool_onrun(ESYS_CONTEXT *context, tpm2_option_flags flags) {

    UNUSED(flags);

    tool_rc rc = init(context);
    if (rc != tool_rc_success) {
        return rc;
    }

    return read_public_and_save(context);
}
